src/main/scala/mesosphere/marathon/upgrade/ReadinessBehavior.scala

```diff
@@ -0,0 +1,122 @@
+package mesosphere.marathon.upgrade
+
+import akka.actor.{ ActorLogging, ActorRef, Actor }
+import mesosphere.marathon.core.readiness.{ ReadinessCheckExecutor, ReadinessCheckResult }
+import mesosphere.marathon.core.task.Task
+import mesosphere.marathon.core.task.tracker.TaskTracker
+import mesosphere.marathon.event.{ DeploymentStatus, HealthStatusChanged, MesosStatusUpdateEvent }
+import mesosphere.marathon.state.{ AppDefinition, PathId, Timestamp }
+import mesosphere.marathon.upgrade.DeploymentManager.ReadinessCheckUpdate
+import mesosphere.marathon.upgrade.ReadinessBehavior.ScheduleReadinessCheckFor
+import rx.lang.scala.Subscription
+
+/**
+  * ReadinessBehavior makes sure all tasks are healthy and ready depending on the app definition.
+  * Listens for TaskStatusUpdate events, HealthCheck events and ReadinessCheck events.
+  * If a task becomes ready, the taskIsReady hook is called.
+  *
+  * Assumptions:
+  *  - the actor is attached to the event stream for HealthStatusChanged and MesosStatusUpdateEvent
+  */
+trait ReadinessBehavior { this: Actor with ActorLogging =>
+
+  import context.dispatcher
+
+  //dependencies
+  def app: AppDefinition
+  def readinessCheckExecutor: ReadinessCheckExecutor
+  def deploymentManager: ActorRef
+  def taskTracker: TaskTracker
+  def status: DeploymentStatus
+
+  //computed values to have stable identifier in pattern matcher
+  val appId: PathId = app.id
+  val version: Timestamp = app.version
+  val versionString: String = version.toString
+
+  //state managed by this behavior
+  var healthy = Set.empty[Task.Id]
+  var ready = Set.empty[Task.Id]
+  var subscriptions = Map.empty[String, Subscription]
+
+  /**
+    * Hook method which is called, whenever a task becomes ready according to the given app definition.
+    * @param taskId the id of the task that has become ready.
+    */
+  def taskIsReady(taskId: Task.Id): Unit
+
+  override def postStop(): Unit = {
+    subscriptions.values.foreach(_.unsubscribe())
+  }
+
+  /**
+    * Depending on the app definition, this method handles:
+    * - app without health checks and without readiness checks
+    * - app with health checks and without readiness checks
+    * - app without health checks and with readiness checks
+    * - app with health checks and with readiness checks
+    *
+    * The #taskIsReady function is called, when the task is ready according to the app definition.
+    */
+  //scalastyle:off cyclomatic.complexity
+  def readinessBehavior: Receive = {
+    def taskIsRunning(taskFn: Task.Id => Unit): Receive = {
+      case MesosStatusUpdateEvent(slaveId, taskId, "TASK_RUNNING", _, `appId`, _, _, _, `versionString`, _, _) => taskFn(taskId) //scalastyle:ignore line.size.limit
```

Comment:
Maybe put `taskFn(taskId)` in a new line instead of adding a scalastyle ignore statement?